pkg/kind/kong_proxy.go

```diff
@@ -0,0 +1,172 @@
+			// fail on timeout
+			err := fmt.Errorf("deployment \"kong-system/ingress-controller-kong\" not ready after %s", time.Minute*3) // TODO: vars here
+			ready <- ProxyReadinessEvent{Err: err}
+			close(ready)
+		}
+	}()
+
+	return
+}
+
+// startDeploymentInformer will watch for the standard Kong proxy Deployment to be posted to the API and will report
+// that deployment to a channel when it is found.
+func (c *kongProxyCluster) startDeploymentInformer(ctx context.Context) (deployment chan *appsv1.Deployment) {
+	deployment = make(chan *appsv1.Deployment)
+	factory := kubeinformers.NewSharedInformerFactory(c.client, time.Minute*3) // TODO: resync vars
+	informer := factory.Apps().V1().Deployments().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			d, ok := obj.(*appsv1.Deployment)
+			if !ok {
+				return
+			}
+			if d.Namespace == "kong-system" && d.Name == "ingress-controller-kong" { // TODO: vars here
+				deployment <- d
+			}
+			return
+		},
+	})
+
+	go func() {
+		informer.Run(ctx.Done())
+		close(deployment)
+	}()
+
+	return
+}
+
+func (c *kongProxyCluster) startServiceInformer(ctx context.Context, nsn types.NamespacedName, ready chan ProxyReadinessEvent) {
+	factory := kubeinformers.NewSharedInformerFactory(c.client, time.Minute*3) // TODO: resync vars
+	informer := factory.Core().V1().Services().Informer()
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		// this UpdateFunc is responsible for closing the ready channel once the proxy URL is ready (or if an error occurs)
+		UpdateFunc: func(old1, new1 interface{}) {
+			newService, ok := new1.(*corev1.Service)
+			if !ok {
+				err := fmt.Errorf("somehow got unexpected type instead of corev1.Service: %T", new1)
+				ready <- ProxyReadinessEvent{Err: err}
+				close(ready)
+				return
+			}
+
+			// if this is the correct service, operate on it
+			if newService.Namespace == nsn.Namespace && newService.Name == nsn.Name {
+				// verify if the IP for the LB has been provisioned yet
+				ing := newService.Status.LoadBalancer.Ingress
+				if len(ing) > 0 && ing[0].IP != "" {
+					// get the URL for the LB
+					u, err := urlForLoadBalancerIngress(newService, &ing[0])
+					if err != nil {
+						ready <- ProxyReadinessEvent{Err: err}
+						close(ready)
+						return
+					}
+
+					// if docker network/metallb test HTTP access using the URL
+					if c.enabledMetal {
+						proxyReady, err := waitForKongProxy(u)
+						if !proxyReady {
+							ready <- ProxyReadinessEvent{Err: err}
+							close(ready)
+							return
+						}
+					}
+
+					ready <- ProxyReadinessEvent{URL: u}
+					close(ready)
+				}
+			}
+		},
+	})
+	go informer.Run(ctx.Done())
+}
+
+func waitForKongProxy(u *url.URL) (proxyReady bool, err error) {
+	timeout := time.Now().Add(time.Minute * 3) // TODO: timeout vars
+	for timeout.After(time.Now()) {
+		var resp *http.Response
+		resp, err = http.Get(u.String())
```

Comment:
I think that this request needs a timeout. Otherwise, this `http.Get` can hang beyond the predefined 3 minutes.
Maybe this could be done by passing `ctx` to `waitForKongProxy` and propagating ctx to the http request, and using `ctx.Done` instead of `timeout.After` 2 lines above here.

Reply:
Sounds good, will change :+1:

Reply:
Fixed in caac2ba4f4a8baa325d5f1334d99f9b82e1125f1